### ISSUE TITLE
hapi: ICatBoxCacheOptions lacks expiresIn, others

### ISSUE THREAD

**Issue by garthk** [DefinitelyTyped/DefinitelyTyped#10060]
- [x] I tried using the latest `hapi/hapi.d.ts` file in this repo and had problems.
  - The authors of that type definition are cc/ @jasonswearingen 

The typings for `ICatBoxCacheOptions` lack almost all the options Hapi supports.

[Hapi API reference: Server Cache Options](https://github.com/hapijs/hapi/blob/v13.5.0/API.md#servercacheoptions):

> Provisions a cache segment within the server cache facility where:
> - `options` - [**catbox** policy](https://github.com/hapijs/catbox#policy) configuration where:
>   - `expiresIn` - relative expiration expressed in the number of milliseconds…
>   - `expiresAt` - time of day expressed in 24h notation using the 'HH:MM' format…
>   - `generateFunc` - a function…
>   - `staleIn` - number of milliseconds…
>   - `staleTimeout` - number of milliseconds…
>   - `generateTimeout` - number of milliseconds…
>   - `generateOnReadError` - if `false`,…
>   - `generateIgnoreWriteError` - if `false`,…
>   - `pendingGenerateTimeout` - number of milliseconds…
>   - `cache` - the cache name configured in [`server.cache`](#server.config.cache)…
>   - `segment` - string segment name…
>   - `shared` - if `true`,…

[`server.method(name, method, options)` _op cit_](https://github.com/hapijs/hapi/blob/v13.5.0/API.md#servermethodname-method-options):

> Methods are registered via `server.method(name, method, [options])` where:
> - `name` - a unique method name used to invoke the method via `server.methods[name]`.
> - `method` - the method function with the signature is one of:
>   - `function(arg1, arg2, ..., argn, next)` where:…
>   - `function(arg1, arg2, ..., argn)` where:…
> - `options` - optional configuration:
>   - `bind` - a context object passed back to the method function (via `this`)…
>   - `cache` - the same cache configuration used in [`server.cache()`](#servercacheoptions)…
>   - `callback` - if `false`…
>   - `generateKey` - a function used to generate a unique key…

[`ICatBoxCacheOptions` in `hapi.d.ts` at `36ce4b6`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/36ce4b6/hapi/hapi.d.ts#L70):

``` typescript
    export interface ICatBoxCacheOptions {
        /**  a prototype function or catbox engine object.  */
        engine: any;
        /**   an identifier used later when provisioning or configuring caching for server methods or plugins. Each cache name must be unique. A single item may omit the name option which defines the default cache. If every cache includes a name, a default memory cache is provisions as well.  */
        name?: string;
        /**   if true, allows multiple cache users to share the same segment (e.g. multiple methods using the same cache storage container). Default to false.  */
        shared?: boolean;
    }
```

[`IServerMethodOptions` _op cit_](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/36ce4b6/hapi/hapi.d.ts#L1084):

``` typescript
    export interface IServerMethodOptions {
        bind?: any;
        cache?: ICatBoxCacheOptions;
        callback?: boolean;
        generateKey?(args: any[]): string;
    }
```

I'll dig in the `joi` schema to nail it down for sure.

**Comment by garthk** [comment]
[`schema.js` at `v13.5.0`](https://github.com/hapijs/hapi/blob/v13.5.0/lib/schema.js):

```js
internals.cachePolicy = Joi.object({
    cache: Joi.string().allow(null).allow(''),
    segment: Joi.string(),
    shared: Joi.boolean()
})
    .options({ allowUnknown: true });               // Catbox validates other keys

// ...

internals.method = Joi.object({
    bind: Joi.object().allow(null),
    generateKey: Joi.func(),
    cache: internals.cachePolicy,
    callback: Joi.boolean()
});
```

**Comment by jasonswearingen** [comment]
hi @garthk the catbox module is newish, please feel free to make a pull request with the required changes, and I'll be happy to accept.

Otherwise, I'll get to this the next time I'm working on hapi related code (but that might be a while)

**Comment by garthk** [comment]
No worries.

**Comment by garthk** [comment]
[`catbox/lib/policy.js` at `v7.1.1`](https://github.com/hapijs/catbox/blob/v7.1.1/lib/policy.js#L279):
```js
internals.schema = Joi.object({
    expiresIn: Joi.number().integer().min(1),
    expiresAt: Joi.string().regex(/^\d\d?\:\d\d$/),
    staleIn: [Joi.number().integer().min(1).max(86400000 - 1), Joi.func()],               // One day - 1 (max is inclusive)
    staleTimeout: Joi.number().integer().min(1),
    generateFunc: Joi.func(),
    generateTimeout: Joi.number().integer().min(1).allow(false),
    generateOnReadError: Joi.boolean(),
    generateIgnoreWriteError: Joi.boolean(),
    dropOnError: Joi.boolean(),
    pendingGenerateTimeout: Joi.number().integer().min(1),

    // Ignored external keys (hapi)

    privacy: Joi.any(),
    cache: Joi.any(),
    segment: Joi.any(),
    shared: Joi.any()
})
```

**Comment by AJamesPhillips** [comment]
We can close this now @andy-ms due to #11974 being merged and #16065 has included these updates and more.

**Comment by AJamesPhillips** [comment]
We can close this now @garthk @DefinitelyTyped due to #11974 being merged and #16065 has included these updates and more.
